docs/main.md

```diff
@@ -0,0 +1,61 @@
+different and is outlined below: 
+
+- **SU2 adapter**
+
+  Since SU2 adapter modifies SU2 source code, it makes no sense to copy clean SU2 build, because it will need to be rebuilt. Instead we just build the adapter
+  from  SU2 source after applying the necessary modifications.
+
+- **dealii adapter**
+  Since dealii is a library and the adapter just needs to link with it, here we actually perform the import of dealii image and only build
+  the actual adapter
+
+- **CalculiX adapter**
+  We only import the necessary libraries (Spooles, ARPACK), since CalculiX adapter modifies CalculiX source.
+
+- **openFOAM adapter**
+  Same as for deal.ii adapter.
+
+- **Nutils adapter**
+  There is no Nutils adapter. There is python script that communicates with Nutils and preCICE. We therefore provide Dockefile for it just for the where it is used,
+  and don't build during adapter building stage, but during test running stage.
+
+- **FEniCS adapter**
+  We install FEniCS, together with adapter using `Dockefile.fenics-adapter` file.
+  
+
+All adapters are built using user `precice` with `gid` and `uid` equal to 1000 and of `precice` group. As a last build stage they create folder for the volume mapping in 
+`/home/precice/Data` folder.
+
+## Running tests
+
+At this stage, we spawn containers with the necessary solvers for the simulation. The execution is controlled using [docker-compose](https://docs.docker.com/compose/). Separate volumes are created for the input and the output of every adapter. Additional docker network is created and used for the communication between containers. Exchange of the data is done using another common volume. The general pattern across of the tests for such volumes to have locations `/home/precice/Data/Exchange`, `/home/precice/Data/Input`,
+`/home/precice/Data/Output`, as explained above.
+
+The tests are based on the corresponding tutorials from the [tutorials repository](https://github.com/precice/tutorials). Several modifications are necessary to decouple the input to solvers, adjust simulation time, or change solver version and to modify `precice-config.xml` for simulation with docker. This is done with `Dockerfile.tutorial_data` image, that clones the tutorial data, adjusts and splits the input to different solvers. It is then is used as first container spawned by `docker-compose`.
+
+Afterwards, when the simulation finishes the output is copied from all the solvers into the common volume and compared to the reference output. An error is thrown if they don't match. Since it is possible for the output of the identical tests to differ in floating points due to specifics of the machine and OS, after simple comparison of the output file by file it is also compared numerically using `compare_results.sh`. It strips aways non-numerical data and then compared files field by field with respect to specified relative tolerance.
```

Comment:
Default tolerance is set in the `compare_results.sh` (it can also adjusted via command line args). The general documentation of all numerical comparison is done there, so I don't duplicate it here.

> Why are we confident that we don't get false successes? 

I did some experiments on my machine with changing the results to be above the tolerance and checking if the error is reported and it is. Of course it is still hard to judge what should be the limit of the tolerance, etc, but given the main cause of failed tests are usually building of adapters of some communication problem and rarely numerics,  I would not put much focus here, until we get more failing cases due to numerics.

Reply:
Usually the building part we already test while developing and it is not the/my main reason we rely on automated testing in this case. We really don't want to introduce numerical changes without noticing, so going conservative in this case would help.

I see that in the `compare_results.sh` we have default values:
```
avg_diff_limit="0.01"
max_diff_limit="0.01"
```
as well as values from arguments (which, if I understand correctly, are not used in `system_testing.py`). So,one needs to know that the tests will complain if the results differ more than 1% (but this default threshold is another discussion).

The information that is missing here (and causes some confusion to an adapter developer) is which values the scripts are used with. So, maybe a clarification would help:

> The default value for the tolerance is set in `compare_results.sh` and can be changed, however the script `system_testing.py` (Travis) uses the default values.